src/microbe_stage/BackgroundPlane.cs

```diff
@@ -0,0 +1,263 @@
+using System;
+using Godot;
+
+/// <summary>
+///   Manages the microbe background plane, optionally applies blur.
+/// </summary>
+public partial class BackgroundPlane : Node3D, IGodotEarlyNodeResolve
+{
+    private readonly StringName blurAmountParameter = new("blurAmount");
+    private readonly StringName textureAlbedoParameter = new("textureAlbedo");
+    private readonly StringName worldPositionParameter = new("worldPos");
+    private readonly StringName lightLevelParameter = new("lightLevel");
+    private readonly StringName distortionStrengthParameter = new("distortionFactor");
+
+    [Export]
+    private NodePath? blurPlanePath;
+
+    [Export]
+    private NodePath blurColorRectPath = null!;
+
+    [Export]
+    private NodePath backgroundPlanePath = null!;
+
+#pragma warning disable CA2213
+
+    /// <summary>
+    ///   Background plane that is moved farther away from the camera when zooming out
+    /// </summary>
+    private Node3D backgroundPlane = null!;
+
+    private Node3D blurPlane = null!;
+
+    private GpuParticles3D? backgroundParticles;
+
+    private ShaderMaterial spatialBlurMaterial = null!;
+
+    private ShaderMaterial canvasBlurMaterial = null!;
+
+    private ShaderMaterial currentBackgroundMaterial = null!;
+
+    private SubViewport subViewport1 = null!;
+
+    private SubViewport subViewport2 = null!;
+#pragma warning restore CA2213
+
+    public bool NodeReferencesResolved { get; private set; }
+
+    public float PlaneOffset
+    {
+        get
+        {
+            return backgroundPlane.Position.Z + 15;
+        }
+        set
+        {
+            backgroundPlane.Position = new Vector3(0, 0, -15 + value);
+        }
+    }
+
+    public override void _Ready()
+    {
+        var material = GetNode<CsgMesh3D>(backgroundPlanePath).Material;
+        var planeBlurMaterial = GetNode<CsgMesh3D>(blurPlanePath).Material;
+        var colorRectBlurMaterial = GetNode<CanvasItem>(blurColorRectPath).Material;
+
+        if (material == null || planeBlurMaterial == null || colorRectBlurMaterial == null)
+        {
+            GD.PrintErr("MicrobeCamera didn't find material to update");
+            return;
+        }
+
+        currentBackgroundMaterial = (ShaderMaterial)material;
+        spatialBlurMaterial = (ShaderMaterial)planeBlurMaterial;
+        canvasBlurMaterial = (ShaderMaterial)colorRectBlurMaterial;
+
+        ResolveNodeReferences();
+
+        ApplyDistortionEffect();
+        ApplyBlurEffect();
+    }
+
+    public void ResolveNodeReferences()
+    {
+        if (NodeReferencesResolved)
+            return;
+
+        NodeReferencesResolved = true;
+
+        if (HasNode(backgroundPlanePath))
+            backgroundPlane = GetNode<Node3D>(backgroundPlanePath);
+
+        if (HasNode(blurPlanePath))
+            blurPlane = GetNode<Node3D>(blurPlanePath);
+
+        if (HasNode("SubViewport"))
+            subViewport1 = GetNode<SubViewport>("SubViewport");
```

Comment:
These should also be done with direct export node references. Also what would be the use case where the subviewports are allowed to be missing?